desc/magnetic_fields.py

```diff
@@ -1803,6 +1810,375 @@ def from_surface(
+            M == 0,
+            jnp.array([N, iota / N, 0, 1 / N]),  # OP
+            jnp.where(
+                N == 0,
+                jnp.array([0, -1, M, -1 / iota]),  # OT
+                jnp.array([N, M * iota / (N - M * iota), M, M / (N - M * iota)]),  # OH
+            ),
+        ).reshape((2, 2))
+
+        # evaluate h(rho,eta,alpha)
+        h = self.shift_basis.evaluate(nodes) @ self.x_lmn + 2 * eta + np.pi
+
+        # solve for (rho,theta_B,zeta_B) corresponding to (rho,eta,alpha)
+        booz = matrix @ np.vstack((alpha, h))
+        return booz[0, :], booz[1, :]
+
+    def compute(
+        self,
+        names,
+        grid=None,
+        params=None,
+        transforms=None,
+        data=None,
+        **kwargs,
+    ):
+        """Compute the quantity given by name on grid.
+
+        Parameters
+        ----------
+        names : str or array-like of str
+            Name(s) of the quantity(s) to compute.
+        grid : Grid or int, optional
+            Grid of coordinates to evaluate at. Defaults to a Linear grid.
+            If an integer, uses that many equally spaced points.
+        params : dict of ndarray
+            Parameters from the equilibrium. Defaults to attributes of self.
+        transforms : dict of Transform
+            Transforms for R, Z, lambda, etc. Default is to build from grid
+        data : dict of ndarray
+            Data computed so far, generally output from other compute functions
+        override_grid : bool
+            If True, override the user supplied grid if necessary and use a full
+            resolution grid to compute quantities and then downsample to user requested
+            grid. If False, uses only the user specified grid, which may lead to
+            inaccurate values for surface or volume averages.
+
+        Returns
+        -------
+        data : dict of ndarray
+            Computed quantity and intermediate variables.
+
+        """
+        if isinstance(names, str):
+            names = [names]
+        if grid is None:
+            grid = LinearGrid(theta=2 * self.M_well, zeta=8, NFP=self.NFP, sym=False)
+        elif not isinstance(grid, _Grid):
+            raise TypeError(
+                "must pass in a Grid object for argument grid!"
+                f" instead got type {type(grid)}"
+            )
+
+        if params is None:
+            params = get_params(names, obj=self)
+        if transforms is None:
+            transforms = get_transforms(names, obj=self, grid=grid, **kwargs)
+        if data is None:
+            data = {}
+        profiles = {}
+
+        data = compute_fun(
+            self,
+            names,
+            params=params,
+            transforms=transforms,
+            profiles=profiles,
+            data=data,
+            **kwargs,
+        )
+        return data
+
+    @property
+    def NFP(self):
+        """int: Number of (toroidal) field periods."""
+        return self._NFP
+
+    @NFP.setter
```

Comment:
the other resolution parameters don't have setter methods, just the `change_resolution` method, any reason for having this?